protocol/x/ratelimit/util/ibc.go

```diff
@@ -0,0 +1,191 @@
+package util
+
+// TODO(CORE-854): Improve attribution message.
+// This file re-uses similar utilities (with minor tweaking) Stride's IBC Rate Limit implementation:
+// https://github.com/Stride-Labs/stride/blob/121f2ac5d2e5f8e406f89999410a49ea4277a552/x/ratelimit
+
+import (
+	"encoding/json"
+	"fmt"
+	"math/big"
+
+	errorsmod "cosmossdk.io/errors"
+	"cosmossdk.io/log"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	ibctransfertypes "github.com/cosmos/ibc-go/v8/modules/apps/transfer/types"
+	channeltypes "github.com/cosmos/ibc-go/v8/modules/core/04-channel/types"
+	"github.com/dydxprotocol/v4-chain/protocol/x/ratelimit/types"
+)
+
+// Parse the denom from the Send Packet that will be used by the rate limit module
+// The denom that the rate limiter will use for a SEND packet depends on whether
+// it was a NATIVE token (e.g. ustrd, stuatom, etc.) or NON-NATIVE token (e.g. ibc/...)...
+//
+// We can identify if the token is native or not by parsing the trace denom from the packet
+// If the token is NATIVE, it will not have a prefix (e.g. ustrd),
+// and if it is NON-NATIVE, it will have a prefix (e.g. transfer/channel-2/uosmo)
+//
+// For NATIVE denoms, return as is (e.g. ustrd)
+// For NON-NATIVE denoms, take the ibc hash (e.g. hash "transfer/channel-2/usoms" into "ibc/...")
+func ParseDenomFromSendPacket(packet ibctransfertypes.FungibleTokenPacketData) (denom string) {
+	// Determine the denom by looking at the denom trace path
+	denomTrace := ibctransfertypes.ParseDenomTrace(packet.Denom)
+
+	// Native assets will have an empty trace path and can be returned as is
+	if denomTrace.Path == "" {
+		denom = packet.Denom
+	} else {
+		// Non-native assets should be hashed
+		denom = denomTrace.IBCDenom()
+	}
+
+	return denom
+}
+
+// Parse the denom from the Recv Packet that will be used by the rate limit module
+// The denom that the rate limiter will use for a RECEIVE packet depends on whether it was a source or sink
+//
+//	Sink:   The token moves forward, to a chain different than its previous hop
+//	        The new port and channel are APPENDED to the denom trace.
+//	        (e.g. A -> B, B is a sink) (e.g. A -> B -> C, C is a sink)
+//
+//	Source: The token moves backwards (i.e. revisits the last chain it was sent from)
+//	        The port and channel are REMOVED from the denom trace - undoing the last hop.
+//	        (e.g. A -> B -> A, A is a source) (e.g. A -> B -> C -> B, B is a source)
+//
+//	If the chain is acting as a SINK: We add on the Stride port and channel and hash it
```

Comment:
nit: did we still want to have `Stride` in the comment? understand this if this is all copy-paste

Reply:
Think it makes more sense to adapt documentation in the context of dYdX chain. Obviously, as part of attribution/licensing we will still make it clear these functions came from Stride implementation